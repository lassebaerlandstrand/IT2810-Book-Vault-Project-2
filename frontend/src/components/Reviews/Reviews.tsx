import { useRef } from 'react';
import { Grid, Group, Stack, Text } from '@mantine/core';
import { Book } from '@/generated/graphql';
import BookReviews from '../BookReviews/BookReviews';
import { Ratings } from '../Ratings/Ratings';
import YourReviewHandler from '../YourReviewHandler/YourReviewHandler';
import styles from './Reviews.module.css';

type ReviewProps = {
  book: Book;
};

const Reviews = ({ book }: ReviewProps) => {
  // For scrolling to top when reaching bottom
  const topOfReviews = useRef<HTMLDivElement>(null);

  return (
    <Group justify="left" gap="lg" ref={topOfReviews}>
      <Stack gap="sm" className={styles.gridWidth}>
        <Grid justify="Space-between">
          <Grid.Col span="auto">
            <Text size="xl" fw={700}>
              Reviews
            </Text>
          </Grid.Col>
          <Grid.Col span="auto">
            <Ratings book={book} justify="right" />
          </Grid.Col>
        </Grid>

        <YourReviewHandler book={book} />

        <BookReviews bookId={book.id} top={topOfReviews} />
      </Stack>
    </Group>
  );
};

export default Reviews;
