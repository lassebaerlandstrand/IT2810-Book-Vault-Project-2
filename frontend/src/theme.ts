import {
  Container,
  createTheme,
  DEFAULT_THEME,
  Loader,
  MantineColorsTuple,
  rem,
} from '@mantine/core';
import { BookLoader } from './components/Loading/Loading';

const CONTAINER_SIZES: Record<string, string> = {
  xxs: '24em',
  xs: '30em',
  sm: '48em',
  md: '64em',
  lg: '75em',
  xl: '90em',
};

// https://colorkit.co/color-palette-generator/02070b-030a11-030d16-051321-06192b-092641-0c3256-0f3f6c-124b81-155896/
const darkBlue: MantineColorsTuple = [
  '#155896',
  '#124b81',
  '#0f3f6c',
  '#0c3256',
  '#092641',
  '#06192b',
  '#051321',
  '#030d16',
  '#030a11',
  '#02070b',
];

// https://colorkit.co/color-palette-generator/6a5f4a-d4bd94-dac5a1-e2d1b5-e9ddc8-eee4d3-f2eade-f6f0e7-f9f5ef-fcfaf7/
const beige: MantineColorsTuple = [
  '#fcfaf7',
  '#f9f5ef',
  '#f6f0e7',
  '#f2eade',
  '#eee4d3',
  '#e9ddc8',
  '#e2d1b5',
  '#dac5a1',
  '#d4bd94',
  '#6a5f4a',
];

export const theme = createTheme({
  fontFamily: `Poppins, ${DEFAULT_THEME.fontFamily}`,
  breakpoints: {
    xxs: '24em', // 384px
    xs: '30em', // 480px
    sm: '48em', // 768px
    md: '64em', // 1024px
    lg: '75em', // 1200px
    xl: '90em', // 1440px
  },
  colors: {
    'dark-blue': darkBlue,
    beige,
  },
  components: {
    Container: Container.extend({
      vars: (_, { size, fluid }) => ({
        root: {
          '--container-size': fluid
            ? '100%'
            : size !== undefined && size in CONTAINER_SIZES
              ? CONTAINER_SIZES[size]
              : rem(size),
        },
      }),
    }),
    Loader: Loader.extend({
      defaultProps: {
        loaders: { ...Loader.defaultLoaders, book: BookLoader },
      },
      vars: (_, props) => {
        if (props.size === 'xxl') {
          return {
            root: {
              '--loader-size': rem(90),
            },
          };
        }
        return { root: {} };
      },
    }),
  },
});
