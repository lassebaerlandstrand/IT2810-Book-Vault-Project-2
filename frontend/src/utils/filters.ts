import { SortBy, SortOrder } from '@/generated/graphql';

export const DEFAULT_FILTERS = {
  sortBy: SortBy.BookName,
  sortOrder: SortOrder.Asc,
  selectedGenres: [],
  selectedPublishers: [],
  selectedAuthors: [],
  selectedBeforeDate: new Date(2020, 11, 31),
  selectedAfterDate: new Date(1900, 0, 1),
  selectedMinPages: undefined,
  selectedMaxPages: undefined,
  selectedMinRating: undefined,
};

export const getFilterParams = (searchParams: URLSearchParams) => {
  const sortBy = (searchParams.get('sortBy') as SortBy) ?? DEFAULT_FILTERS.sortBy;
  const sortOrder = (searchParams.get('sortOrder') as SortOrder) ?? DEFAULT_FILTERS.sortOrder;
  const authors = searchParams.getAll('authors') ?? DEFAULT_FILTERS.selectedAuthors;
  const publishers = searchParams.getAll('publishers') ?? DEFAULT_FILTERS.selectedPublishers;
  const genres = searchParams.getAll('genres') ?? DEFAULT_FILTERS.selectedGenres;
  const beforeDate = searchParams.get('beforeDate')
    ? new Date(searchParams.get('beforeDate') ?? '')
    : DEFAULT_FILTERS.selectedBeforeDate;
  const afterDate = searchParams.get('afterDate')
    ? new Date(searchParams.get('afterDate') ?? '')
    : DEFAULT_FILTERS.selectedAfterDate;
  const minPages =
    parseInt(searchParams.get('minPages') ?? '', 10) || DEFAULT_FILTERS.selectedMinPages;
  const maxPages =
    parseInt(searchParams.get('maxPages') ?? '', 10) || DEFAULT_FILTERS.selectedMaxPages;
  const minRating =
    parseInt(searchParams.get('minRating') ?? '', 10) || DEFAULT_FILTERS.selectedMinRating;

  return {
    sortBy,
    sortOrder,
    authors,
    publishers,
    genres,
    beforeDate,
    afterDate,
    minPages,
    maxPages,
    minRating,
  };
};
