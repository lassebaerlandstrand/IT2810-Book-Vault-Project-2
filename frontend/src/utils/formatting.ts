/**
 * Formats a number by adding spaces as thousand separators.
 *
 * @param {number} number - The input number to format
 * @returns {string} The formatted number string with space separators
 *
 * @example
 * // returns '1 000'
 * formatNumberWithSpaces(1000);
 *
 * @example
 * // returns '1 000 000'
 * formatNumberWithSpaces(1000000);
 */
export const formatNumberWithSpaces = (number: number) =>
  number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ' ');

// The regular expression used in the replace method:
// \B - matches a non-word boundary to avoid adding spaces at the start of the string.
// (?=(\d{3})+(?!\d)) - positive lookahead to find positions followed by a group of three digits not followed by another digit.
