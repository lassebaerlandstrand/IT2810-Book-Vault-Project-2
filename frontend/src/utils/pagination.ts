const DEFAULT_PAGE = 1;
const DEFAULT_LIMIT = 25;
const LIMIT_OPTIONS = ['10', '25', '50', '100'];

export const getPaginationParams = (searchParams: URLSearchParams) => {
  const page = Number(searchParams.get('page')) || DEFAULT_PAGE;
  const limit = Number(searchParams.get('limit')) || DEFAULT_LIMIT;

  return { page, limit, DEFAULT_PAGE, DEFAULT_LIMIT, LIMIT_OPTIONS };
};
