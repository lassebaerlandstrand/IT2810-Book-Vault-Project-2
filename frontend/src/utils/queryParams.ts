import { SetURLSearchParams } from 'react-router-dom';

/**
 * Updates the URL search parameters based on the provided key and value.
 * Handles different types of values (single string, array of strings, or undefined).
 * This function handles consecutive calls.
 *
 * @param {SetURLSearchParams} setSearchParams - Function to update the URL search parameters
 * @param {string} key - The key of the query parameter to update
 * @param {string | string[] | undefined} value - The value to set for the query parameter
 *
 * @example
 * updateQueryParams(setSearchParams, 'search', 'newSearchQuery')
 * updateQueryParams(setSearchParams, 'genres', ['tag1', 'tag2'])
 * updateQueryParams(setSearchParams, 'page', '5')
 * updateQueryParams(setSearchParams, 'page', undefined) // Removes the key
 *
 */
export const updateQueryParams = (
  setSearchParams: SetURLSearchParams,
  key: string,
  value: string | string[] | undefined
) => {
  setSearchParams((params) => {
    if (Array.isArray(value)) {
      params.delete(key);
      value.forEach((v) => params.append(key, v));
    } else if (value === undefined || value === '') {
      params.delete(key);
    } else {
      params.set(key, value);
    }
    return params;
  });
};
