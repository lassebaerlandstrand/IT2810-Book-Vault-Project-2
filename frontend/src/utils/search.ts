/**
 * Extracts the search parameter from the URL.
 * This function ensures that the search parameters are correctly parsed and returned in a consistent format.
 *
 * @param {URLSearchParams} searchParams - URL search parameters containing search criteria
 * @returns {Object} Search parameters
 *
 */
export const getSearchParams = (searchParams: URLSearchParams) => {
  const searchValue = searchParams.get('search') || '';
  return { searchValue };
};
