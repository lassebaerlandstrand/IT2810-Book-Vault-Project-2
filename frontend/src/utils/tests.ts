/**
 * Removes random attributes generated by Mantine from DOM elements.
 * This is necessary to ensure consistent snapshots in tests, as Mantine uses random IDs.
 */
export const removeMantineRandomAttributes = () => {
  const attributesToRemove = [
    ...document.body.querySelectorAll('div [id^="mantine"]'),
    ...document.body.querySelectorAll('div [for^="mantine"]'),
  ];
  attributesToRemove.forEach((element) => {
    element.removeAttribute('for');
    element.removeAttribute('id');
    element.removeAttribute('class');
    element.removeAttribute('aria-describedby');
  });
};
