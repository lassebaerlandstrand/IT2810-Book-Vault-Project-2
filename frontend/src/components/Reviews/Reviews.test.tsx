import { MockedProvider } from '@apollo/client/testing';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { MantineProvider } from '@mantine/core';
import { theme } from '@/theme';
import { removeMantineRandomAttributes } from '@/utils/tests';
import { dummyBook } from '../../../test-utils/testVars';
import Reviews from './Reviews';

vi.mock('recharts', async (importOriginal) => {
  const originalModule = (await importOriginal()) as Record<string, unknown>;
  return {
    ...originalModule,
    ResponsiveContainer: ({ children }: { children: React.ReactNode }) => (
      <div style={{ width: '100%', height: '100%' }}>{children}</div>
    ),
  };
});

const customRender = () => {
  return render(
    <MockedProvider addTypename={false}>
      <MantineProvider theme={theme}>
        <MemoryRouter>
          <Reviews book={dummyBook} />
        </MemoryRouter>
      </MantineProvider>
    </MockedProvider>
  );
};

describe('Reviews Component', () => {
  beforeEach(() => {
    // Mocking BookReviews and YourReviewHandler as these have their own tests

    vi.mock('@/components/BookReviews/BookReviews', () => ({
      default: () => {
        return <div>BookReview mock</div>;
      },
    }));

    vi.mock('@/components/YourReviewHandler/YourReviewHandler', () => ({
      default: () => {
        return <div>YourReviewHandler mock</div>;
      },
    }));
  });

  test('header and average rating renders properly', async () => {
    customRender();

    expect(screen.getByText('Reviews')).toBeInTheDocument();
    expect(screen.getByText(dummyBook.rating.toFixed(1))).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = customRender();

    removeMantineRandomAttributes();
    expect(asFragment()).toMatchSnapshot();
  });
});
