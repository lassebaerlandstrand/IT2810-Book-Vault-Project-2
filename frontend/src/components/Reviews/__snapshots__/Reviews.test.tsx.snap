// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Reviews Component > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-logo-theme-0: #fefefe;--mantine-color-logo-theme-1: #cbcfd2;--mantine-color-logo-theme-2: #97a0a6;--mantine-color-logo-theme-3: #7d8991;--mantine-color-logo-theme-4: #63717b;--mantine-color-logo-theme-5: #495a65;--mantine-color-logo-theme-6: #2f434f;--mantine-color-logo-theme-7: #132937;--mantine-color-logo-theme-8: #0b202e;--mantine-color-logo-theme-9: #061017;}:root[data-mantine-color-scheme="dark"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-8);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-9);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.15);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.2);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-3);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-outline-hover: rgba(99, 113, 123, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-filled);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-7);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.1);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.12);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline-hover: rgba(47, 67, 79, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="m_4081bf90 mantine-Group-root"
    style="--group-gap: var(--mantine-spacing-lg); --group-align: center; --group-justify: left; --group-wrap: wrap;"
  >
    <div
      class="_gridWidth_8ec61a m_6d731127 mantine-Stack-root"
      style="--stack-gap: var(--mantine-spacing-sm); --stack-align: stretch; --stack-justify: flex-start;"
    >
      <style
        data-mantine-styles="inline"
      >
        .__m__-r17{--grid-gutter:var(--mantine-spacing-md);}
      </style>
      <div
        class="m_410352e9 mantine-Grid-root __m__-r17"
        style="--grid-justify: Space-between;"
      >
        <div
          class="m_dee7bd2f mantine-Grid-inner"
        >
          <style
            data-mantine-styles="inline"
          >
            .__m__-r19{--col-flex-grow:1;--col-flex-basis:0rem;--col-max-width:100%;}
          </style>
          <div
            class="m_96bdd299 mantine-Grid-col __m__-r19"
          >
            <p
              class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
              data-size="xl"
              style="--text-fz: var(--mantine-font-size-xl); --text-lh: var(--mantine-line-height-xl); font-weight: 700;"
            >
              Reviews
            </p>
          </div>
          <style
            data-mantine-styles="inline"
          >
            .__m__-r1c{--col-flex-grow:1;--col-flex-basis:0rem;--col-max-width:100%;}
          </style>
          <div
            class="m_96bdd299 mantine-Grid-col __m__-r1c"
          >
            <div
              class="m_8bffd616 mantine-Flex-root __m__-r1e"
              style="gap: calc(0.4375rem * var(--mantine-scale)); align-items: center; justify-content: right;"
            >
              <div
                data-size="sm"
                style="--rating-size: var(--rating-size-sm); --rating-color: var(--mantine-color-yellow-filled);"
              >
                <div
                  class="m_61734bb7 mantine-Rating-symbolGroup"
                >
                  <div
                    data-read-only="true"
                    style="--rating-item-z-index: 0;"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                      style="--rating-symbol-clip-path: inset(0 100% 0 0);"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    data-read-only="true"
                    style="--rating-item-z-index: 0;"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                      style="--rating-symbol-clip-path: inset(0 100% 0 0);"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    data-read-only="true"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div
                  class="m_61734bb7 mantine-Rating-symbolGroup"
                >
                  <div
                    data-read-only="true"
                    style="--rating-item-z-index: 0;"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                      style="--rating-symbol-clip-path: inset(0 100% 0 0);"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    data-read-only="true"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div
                  class="m_61734bb7 mantine-Rating-symbolGroup"
                >
                  <div
                    data-read-only="true"
                    style="--rating-item-z-index: 0;"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                      style="--rating-symbol-clip-path: inset(0 100% 0 0);"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    data-read-only="true"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div
                  class="m_61734bb7 mantine-Rating-symbolGroup"
                >
                  <div
                    data-read-only="true"
                    style="--rating-item-z-index: 2;"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                      style="--rating-symbol-clip-path: inset(0 50% 0 0);"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        data-filled="true"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    data-read-only="true"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div
                  class="m_61734bb7 mantine-Rating-symbolGroup"
                >
                  <div
                    data-read-only="true"
                    style="--rating-item-z-index: 0;"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                      style="--rating-symbol-clip-path: inset(0 100% 0 0);"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    data-read-only="true"
                  >
                    <div
                      class="m_fae05d6a mantine-Rating-symbolBody"
                    >
                      <svg
                        class="m_5662a89a mantine-Rating-starSymbol"
                        fill="none"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 17.75l-6.172 3.245l1.179 -6.873l-5 -4.867l6.9 -1l3.086 -6.253l3.086 6.253l6.9 1l-5 4.867l1.179 6.873z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
              <p
                class="mantine-focus-auto m_b6d8b162 mantine-Text-root"
                style="font-weight: 500;"
              >
                3.6
              </p>
            </div>
          </div>
        </div>
      </div>
      <div>
        YourReviewHandler mock
      </div>
      <div>
        BookReview mock
      </div>
    </div>
  </div>
</DocumentFragment>
`;
